source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -4650,6 +4751,221 @@ export class ChoiceSetInput extends Input {
+        choiceSetContainer.style.width = "100%";
+
+        this._textInput = document.createElement("input");
+        this._textInput.className = this.hostConfig.makeCssClassName(
+            "ac-input",
+            "ac-multichoiceInput",
+            "ac-choiceSetInput-filtered"
+        );
+        this._textInput.type = "text";
+        this._textInput.style.width = "100%";
+
+        this._textInput.onkeydown = (event) => {
+            if (event.key === "ArrowDown") {
+                const firstChoice = document.getElementById("ac-choiceSetInput-choice-0");
+                if (firstChoice) {
+                    firstChoice.focus();
+                }
+            }
+        };
+
+        this._dropdown = document.createElement("div");
+        this._dropdown.className = this.hostConfig.makeCssClassName(
+            "ac-input",
+            "ac-multichoiceInput",
+            "ac-choiceSetInput-filtered-dropdown"
+        );
+
+        choiceSetContainer.append(this._textInput, this._dropdown);
+
+        document.onclick = (event) => {
+            if (this._dropdown) {
+                let child = this._dropdown.firstChild;
+                while (child && event.target !== child) {
+                    child = child.nextSibling;
+                }
+                // Dropdown closes if user clicks outside the choiceset.
+                if (child || !(event.target === this._textInput)) {
+                    this._dropdown.style.display = "none";
+                }
+            }
+        };
+        this._renderedElement = choiceSetContainer;
+    }
+
+    createChoice(value: string, id: number): HTMLSpanElement {
+        const choice = document.createElement("span");
+        choice.className = this.hostConfig.makeCssClassName("ac-input", "ac-choiceSetInput-choice");
+        choice.id = `ac-choiceSetInput-choice-${id}`;
+        choice.innerText = value;
+        choice.tabIndex = -1;
+
+        choice.onclick = () => {
+            if (this._textInput) {
+                this._textInput.value = choice.innerText;
+            }
+            if (this._dropdown) {
+                this._dropdown.style.display = "none";
+            }
+        };
+        choice.onkeydown = (event) => {
+            if (event.key === "ArrowDown") {
+                const nextChoice = document.getElementById(`ac-choiceSetInput-choice-${id + 1}`);
+                if (nextChoice) {
+                    nextChoice.focus();
+                }
+            } else if (event.key === "ArrowUp") {
+                const prevChoice = document.getElementById(`ac-choiceSetInput-choice-${id - 1}`);
+                if (prevChoice) {
+                    prevChoice.focus();
+                }
+            } else if (event.key === "Enter") {
+                choice.click();
+            }
+        };
+        return choice;
+    }
+
+    filterChoices(isDynamic?: boolean) {
+        const filter = this._textInput?.value.toLowerCase() || "";
```

Comment:
Can the card author provide static and dynamic choices? If so, would we want to filter both arrays here?

Reply:
We filter the static choices and dynamic choices at separate times. Notice that we call the filterChoices method the first time (for filtering only the static choices) in the flow when we show the loading indicator. This is done because we want to show the static choices as soon as possible to the user. Dynamic choices will take time to be fetched via the network call. And when they do come, we will filter and add them to the list. It is for a better UX, hence the design.